### PR TITLE
refactor(client): remove goroutine synchronization in DoDeadline

### PR DESCRIPTION
**Pull request by Duslia**
#### What type of PR is this?

refactor
#### Check the PR title.

- [ ] This PR title match the format: \<type\>(optional scope): \<description\>
- [ ] The description of this PR title is user-oriented and clear enough for others to understand.

#### (Optional) Translate the PR title into Chinese.
去除 DoDeadline 中的 goroutine 同步

#### (Optional) More detail description for this PR(en: English/zh: Chinese).

en: Refactor client.DoDeadline, control the timeout through dialtimeout, writetimeout, readtimeout, reduce the number of goroutines and copies
zh(optional): 重构 client.DoDeadline，通过 dialtimeout、writetimeout、readtimeout 控制超时，减少协程和拷贝数量

### CODE REVIEW

File: pkg/protocol/request.go [hunk 1]
```diff
@@ -113,6 +114,10 @@ type Request struct {
 
 	// Request level options, service discovery options etc.
 	options *config.RequestOptions
+
+	// Request timeout. Usually set by DoDeadline or DoTimeout
+	// if <= 0, means not set
+	timeout time.Duration
```
Comment (welkeyever) [inline pkg/protocol/request.go:120]: 请求级别的option？统一放到options里？

File: pkg/protocol/http1/client.go [hunk 2]
```diff
@@ -422,14 +428,14 @@ func (c *HostClient) PendingRequests() int {
 	return int(atomic.LoadInt32(&c.pendingRequests))
 }
 
-func (c *HostClient) do(req *protocol.Request, resp *protocol.Response) (bool, error) {
+func (c *HostClient) do(req *protocol.Request, resp *protocol.Response, before time.Time) (bool, error) {
```
Comment (welkeyever) [inline pkg/protocol/http1/client.go:431]: before作为请求级别超时记录辅助变量可以一起放到请求级别option里？这样应该就不需要动这个函数签名了

File: pkg/protocol/client/client.go [hunk 3]
```diff
@@ -279,76 +277,20 @@ func getRedirectURL(baseURL string, location []byte) string {
 }
 
 func DoTimeout(ctx context.Context, req *protocol.Request, resp *protocol.Response, timeout time.Duration, c Doer) error {
-	deadline := time.Now().Add(timeout)
-	return DoDeadline(ctx, req, resp, deadline, c)
+	if timeout <= 0 {
+		return errTimeout
+	}
+	// Note: it will overwrite the reqTimeout.
+	req.SetOptions(config.WithReqTimeout(timeout))
```
Comment (welkeyever) [inline pkg/protocol/client/client.go:284]: 可以增量更新吗

Comment (Duslia) [reply pkg/protocol/client/client.go:284]: yes

File: pkg/common/config/request_option.go [hunk 4]
```diff
@@ -95,6 +99,17 @@ func WithWriteTimeout(t time.Duration) RequestOption {
 	}}
 }
 
+// WithReqTimeout sets whole request timeout. If it reaches timeout,
+// the client will return.
+//
+// This is the request level configuration.
+func WithReqTimeout(t time.Duration) RequestOption {
+	return RequestOption{F: func(o *RequestOptions) {
+		o.begin = time.Now()
```
Comment (welkeyever) [inline pkg/common/config/request_option.go:108]: with的时候不一定是请求发起的时候req数据准备如果和发起请求之间有较长时间，应该超时就对不上了，这个时间最好在Do里面更新吧

**Comment by Duslia** [comment]
benchamrk: 
previous version
image
improve version: 
HKMozTnkm8

File: pkg/protocol/client/client.go [hunk 5]
```diff
@@ -279,76 +277,20 @@ func getRedirectURL(baseURL string, location []byte) string {
 }
 
 func DoTimeout(ctx context.Context, req *protocol.Request, resp *protocol.Response, timeout time.Duration, c Doer) error {
-	deadline := time.Now().Add(timeout)
-	return DoDeadline(ctx, req, resp, deadline, c)
+	if timeout <= 0 {
+		return errTimeout
+	}
+	// Note: it will overwrite the reqTimeout.
+	req.SetOptions(config.WithRequestTimeout(timeout), config.WithBegin(time.Now()))
```
Comment (welkeyever) [inline pkg/protocol/client/client.go:284]: 增量更新貌似没做到，比如提前配置在req上的option

File: pkg/protocol/http1/client.go [hunk 6]
```diff
@@ -566,14 +597,18 @@ func (c *HostClient) doNonNilReqResp(req *protocol.Request, resp *protocol.Respo
...
-		// See https://github.com/golang/go/issues/15133#issuecomment-271571395 for details
-		if err = conn.SetReadTimeout(rc.readTimeout); err != nil {
-			c.closeConn(cc)
-			// try another connection if retry is enabled
-			return true, err
-		}
+	shouldClose, timeout = updateReqTimeout(reqTimeout, rc.readTimeout, begin)
+	if shouldClose {
+		c.closeConn(cc)
+		return true, errTimeout
```
[the remark this answers is not in the document]
Comment (welkeyever) [reply pkg/protocol/http1/client.go]: cc https://github.com/cloudwego/hertz/pull/681/files#diff-6ca4c04e31a8d5a96acc20e1ec8eb8462c1d50d6388a4af863a2cd5bf33327b1R102

**Review by welkeyever — APPROVED** [review]

**Review by FGYFFFF — APPROVED** [review]

**Pull request merged**